Add Fake Tensor deepcopy impl

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79581
* __->__ #79580

So, the goal here is to take an `nn.Module` and convert all of its tensors to `FakeTensors`, and with the correct mapping from the original Tensors to their new FakeTensors which is needed for consistent storages/aliasing relationships in the new FakeTensors. 

A couple things that don't work:
- defining `__torch_function__` for `__deepcopy__` on FakeTensors because the original Tensors were not FakeTensors to begin with. 
- Deepcopying the module and then converting each Tensor attribute to FakeTensor manually. This doesn't set up the right mapping from original Tensor -> new Fake Tensor in case any of those Tensor attributes are used elsewhere.
- Maybe in the future making `FakeTensorMode` also inherit from `TorchFunctionMode` and defining a `__torch__function__` for deepcopy might work, but at present there isnt infrastructure for inheriting from both TorchFunction and TorchDispatch, and when I tried a trivial version of overriding deepcopy with a mode I ran into errors.


Edit: no longer accepting non-fake inputs, and introducing a TorchFunctionMode just for `__deepcopy__`, and invoking it in torch.Tensor/torch.nn.Parameter

The changes here are minimal and we can switch to the third in the future as needed.

Also updates the dynamic shape checking to use tags.

eellison committed:
Add option for allowing non-fake inputs, add deepcopy impl

[ghstack-poisoned]

eellison committed:
Update on "Add option for allowing non-fake inputs, add deepcopy impl"




[ghstack-poisoned]

eellison committed:
Update on "Add option for allowing non-fake inputs, add deepcopy impl"




[ghstack-poisoned]

eellison committed:
Update on "Add option for allowing non-fake inputs, add deepcopy impl"


So, the goal here is to take an `nn.Module` and convert all of its tensors to `FakeTensors`, and with the correct mapping from the original Tensors to their new FakeTensors which is needed for consistent storages/aliasing relationships in the new FakeTensors. 

A couple things that don't work:
- defining `__torch_function__` for `__deepcopy__` on FakeTensors because the original Tensors were not FakeTensors to begin with. 
- Deepcopying the module and then converting each Tensor attribute to FakeTensor manually. This doesn't set up the right mapping from original Tensor -> new Fake Tensor in case any of those Tensor attributes are used elsewhere.
- Maybe in the future making `FakeTensorMode` also inherit from `TorchFunctionMode` and defining a `__torch__function__` for deepcopy might work, but at present there isnt infrastructure for inheriting from both TorchFunction and TorchDispatch, and when I tried a trivial version of overriding deepcopy with a mode I ran into errors.

The changes here are minimal and we can switch to the third in the future as needed.

Also updates the dynamic shape checking to use tags.


[ghstack-poisoned]

samdow commented on torch/_tensor.py:
```diff
@@ -96,7 +96,14 @@ def __deepcopy__(self, memo):
             # Update the test in test_serialization if you remove 'meta' from here
-            if self.is_sparse or self.device.type in ['lazy', 'xla', 'mps', 'ort', 'meta', 'hpu'] or \
+
+            fake_tensor_dispatch = isinstance(
```
Could we try something like:
```
class FakeTensorMode:
  def __init__(...):
    self.dispatch_mode = FakeTensorDispatchMode()  # the current FakeTensorMode
    self.function_mode = FakeTensorFunctionMode()  # one that only defines a torch_function
  def __enter__(...):
     with self.dispatch_mode:
       with self.function_mode:
          yield
```

then use this like `with FakeTensorMode()` (this is a new API that we're trying to switch over to lol sorry for all the changes)

eellison replied:
When I try just a pretty small self-contained version I get an error:
```
import torch
import copy
from torch.overrides import TorchFunctionMode


class ModeTensor(torch.Tensor):
    def __new__(cls, elem, mode):
        r = torch.Tensor._make_subclass(cls, elem, elem.requires_grad)
        r.elem = elem
        r.mode = mode
        return r

    def __torch_function__(self, func, types, args=(), kwargs=None):
        with self.mode:
            return func(*args, **kwargs)

class Mode(TorchFunctionMode):
    def __torch_function__(self, func, types, args=(), kwargs=None):
        def unwrap(e):
            if isinstance(e, ModeTensor):
                return e.elem
            else:
                return e

        def wrap(t):
            if isinstance(t, torch.Tensor):
                return ModeTensor(t, self)
            else:
                return t

        kwargs = kwargs if kwargs else {}
        return wrap(func(*tuple(unwrap(a) for a in args), **kwargs))

x = torch.rand([4])
mod = torch.nn.Linear(10, 10)
with Mode():
    mod2 = copy.deepcopy(mod)
print(mod2)
print(mod2.weight)
```

> Traceback (most recent call last):
  File "/scratch/eellison/elias.py", line 39, in <module>
    print(mod2.weight)
  File "/private/home/eellison/anaconda3/lib/python3.8/site-packages/torch/_tensor.py", line 350, in __repr__
    return handle_torch_function(Tensor.__repr__, (self,), self,
  File "/private/home/eellison/anaconda3/lib/python3.8/site-packages/torch/overrides.py", line 1522, in handle_torch_function
    result = torch_func_method(public_api, types, args, kwargs)
  File "/scratch/eellison/elias.py", line 14, in __torch_function__
    with self.mode:
  File "/private/home/eellison/anaconda3/lib/python3.8/site-packages/torch/overrides.py", line 1847, in __enter__
    raise RuntimeError(f"{self} has already been used as a mode. Please use a fresh version or use restore")
RuntimeError: <__main__.Mode object at 0x7fc0db542820> has already been used as a mode. Please use a fresh version or use restore

eellison replied:
(same thing regardless if I use `inner=None` or `with mode.push()`

eellison replied:
The changes here are extremely minimal (~5 lines) and I would rather not block on fixing internal TorchDispatch errors or figuring the inheritance of both TorchFunctionMode and TorchDispatchMode which doesnt exist today, especially since the APIs are still in flux.

samdow replied:
Ahh--could you try that but `with self.mode.restore()`

samdow replied:
> The changes here are extremely minimal (~5 lines) and I would rather not block on fixing internal TorchDispatch errors or figuring the inheritance of both TorchFunctionMode and TorchDispatchMode which doesnt exist today, especially since the APIs are still in flux.

I think the idea of leaking details about the existence of mode, let alone the specific FakeTensor mode, to tensor internals is pretty bad practice* and is directly backwards of how we generally want the relationship of modes and functions (modes should know about functions, not the other way around). If there's any way that we can make this work without changing the tensor implementation itself, it seems preferable

* Specifically if someone had a limited build that excludes fake_tensor.py, they can no longer use pytorch at all

eellison replied:
```
Traceback (most recent call last):
  File "/scratch/eellison/elias.py", line 40, in <module>
    print(mod2.weight)
  File "/private/home/eellison/anaconda3/lib/python3.8/site-packages/torch/_tensor.py", line 350, in __repr__
    return handle_torch_function(Tensor.__repr__, (self,), self,
  File "/private/home/eellison/anaconda3/lib/python3.8/site-packages/torch/overrides.py", line 1522, in handle_torch_function
    result = torch_func_method(public_api, types, args, kwargs)
  File "/scratch/eellison/elias.py", line 14, in __torch_function__
    with self.mode.restore():
  File "/private/home/eellison/anaconda3/lib/python3.8/contextlib.py", line 113, in __enter__
    return next(self.gen)
  File "/private/home/eellison/anaconda3/lib/python3.8/site-packages/torch/utils/_mode_utils.py", line 96, in _restore_mode
    raise RuntimeError(f"{mode} does not have any ancestors. Use the standard version instead of restore")
RuntimeError: <__main__.Mode object at 0x7f5359fc5f40> does not have any ancestors. Use the standard version instead of restore
```

eellison replied:
I will try to introduce a separate Mode class for deepcopy, although I think the api is a bit wonky.

samdow replied:
I agree it's wonky and generally the concept of having some interception at the torch function and other interception at the torch dispatch level isn't something we'd considered. Let me know if you need help writing it. Also JFYI that I wasn't able to repro the second error and I'm not certain what's up with that

eellison replied:
So, these are the functions that get called if you call deepcopy are:
```
<method-wrapper '__get__' of getset_descriptor object at 0x7efb77191f80>
<method 'clone' of 'torch._C._TensorBase' objects>
<function Tensor.__hash__ at 0x7efb76fda3a0>
<method-wrapper '__get__' of getset_descriptor object at 0x7efb77193280>
<method 'detach' of 'torch._C._TensorBase' objects>
<method 'requires_grad_' of 'torch._C._TensorBase' objects>
<method-wrapper '__get__' of getset_descriptor object at 0x7efb77191f80>
<method 'clone' of 'torch._C._TensorBase' objects>
<function Tensor.__hash__ at 0x7efb76fda3a0>
<method-wrapper '__get__' of getset_descriptor object at 0x7efb77193280>
<method 'detach' of 'torch._C._TensorBase' objects>
<method 'requires_grad_' of 'torch._C._TensorBase' objects>
<method-wrapper '__get__' of getset_descriptor object at 0x7efb77191f80>
<method 'clone' of 'torch._C._TensorBase' objects>
<function Tensor.__hash__ at 0x7efb76fda3a0>
<method-wrapper '__get__' of getset_descriptor object at 0x7efb77193280>
<method 'detach' of 'torch._C._TensorBase' objects>
<method 'requires_grad_' of 'torch._C._TensorBase' objects>
```

The following seems to work:
```
class Mode(TorchFunctionMode):
    def __init__(self, fake_mode):
        self.fake_mode = fake_mode

    def __torch_function__(self, func, types, args=(), kwargs=None):
        def wrap(t):
            if isinstance(t, torch.Tensor) and not isinstance(t, FakeTensor):
                return self.fake_mode.from_tensor(t)
            else:
                return t

        kwargs = kwargs if kwargs else {}
        return wrap(func(*args, **kwargs))
```

but I'm not convinced it's 100p correct.. either way going with that for now I guess

eellison replied:
Hmm this doesnt copy over `.grad`...  😢

samdow replied:
In the mode, can we just check if the func is `Tensor.__deepcopy__` and then have that just return `self.fake_mode.from_tensor(self)` and not call the function again since that's the behavior that's it looks like was originally here

eellison replied:
The original tensor is not FakeTensor, so `Tensor.__deepcopy__` does not get called. I think I need to add a check if a `TorchFunctionMode` is active here https://github.com/pytorch/pytorch/blob/master/torch/_tensor.py#L86 and call `__deepcopy__` on the mode with tensors

samdow replied:
`has_torch_function_unary ` [has a check that the mode is on](https://github.com/pytorch/pytorch/blob/master/torch/csrc/utils/disable_torch_function.cpp#L223)...really weirdly, with some debugging, I'm not even seeing `__deepcopy__` being called in this case (if I add a breakpoint at the top of that function, it never gets hit)

samdow replied:
Okay so what's happening is that [Parameter has a custom deepcopy](https://github.com/pytorch/pytorch/blob/master/torch/nn/parameter.py#L52) that's being called and it doesn't redispatch to tensor's deepcopy. I'm a little confused how it was working before with a change to `_tensor.py`?

To double check, the main use case for this is in fact modules?

If so, a couple thoughts:
- Do we want to always continue tracking gradients through clone? If so, we should be able to have special behavior for clone in the dispatch mode that doesn't redispatch to `.clone()` and just uses the same underlying tensor
- Otherwise, it might be okay to add a funky `check_torch_function` to parameter's deep copy that does like
```
if _get_torch_function_mode():
  res_data = handle_torch_function('Tensor.__deep_copy__', ...)
else:
  res_data = self.data.clone(memory_format=torch.preserve_format)
result = type(self)(res_data, self.requires_grad)
```

However, if we go with the second option, I would really want to know why we want to use the same tensor in the deep copy case and not in other cases.

cc @zou3519 for any other brainstorming ideas/vetos of current ideas

eellison replied:
In my current diff: I added
```
if _get_torch_function_mode():
      return handle_torch_function(Parameter.__deepcopy__, (self,), self, memo)
```` 

which is pretty similar

samdow replied:
If we do this, note that we will now be returning a FakeTensor(Parameter( instead of a Parameter(FakeTensor()) and it looks like the tests [are expecting different behavior](https://github.com/pytorch/pytorch/pull/79580/files#diff-3bc0cf540b694f4ec0a3749f78b047456657a53a5657e495ffb68e5970c5fdaaR167). Also I'm still unclear why we want different behavior for deep copy than with normal clone

eellison replied:
> note that we will now be returning a FakeTensor(Parameter( instead of a Parameter(FakeTensor())

I think that's dependent on the implementation - that's now it's working rn afaik.

eellison committed:
Update on "Add option for allowing non-fake inputs, add deepcopy impl"


So, the goal here is to take an `nn.Module` and convert all of its tensors to `FakeTensors`, and with the correct mapping from the original Tensors to their new FakeTensors which is needed for consistent storages/aliasing relationships in the new FakeTensors. 

A couple things that don't work:
- defining `__torch_function__` for `__deepcopy__` on FakeTensors because the original Tensors were not FakeTensors to begin with. 
- Deepcopying the module and then converting each Tensor attribute to FakeTensor manually. This doesn't set up the right mapping from original Tensor -> new Fake Tensor in case any of those Tensor attributes are used elsewhere.
- Maybe in the future making `FakeTensorMode` also inherit from `TorchFunctionMode` and defining a `__torch__function__` for deepcopy might work, but at present there isnt infrastructure for inheriting from both TorchFunction and TorchDispatch, and when I tried a trivial version of overriding deepcopy with a mode I ran into errors.

The changes here are minimal and we can switch to the third in the future as needed.

Also updates the dynamic shape checking to use tags.


[ghstack-poisoned]

Chillee commented on torch/_tensor.py:
```diff
@@ -96,7 +96,14 @@ def __deepcopy__(self, memo):
             # Update the test in test_serialization if you remove 'meta' from here
-            if self.is_sparse or self.device.type in ['lazy', 'xla', 'mps', 'ort', 'meta', 'hpu'] or \
+
+            fake_tensor_dispatch = isinstance(
```
Can we just override `__deepcopy__` on the FakeTensor instead of modifying `__deepcopy__` on torch.Tensor itself?

eellison replied:
They're not FakeTensors at this point, we're deepcopying the original regular Tensors to FakeTensors.

eellison committed:
Update on "Add option for allowing non-fake inputs, add deepcopy impl"


So, the goal here is to take an `nn.Module` and convert all of its tensors to `FakeTensors`, and with the correct mapping from the original Tensors to their new FakeTensors which is needed for consistent storages/aliasing relationships in the new FakeTensors. 

A couple things that don't work:
- defining `__torch_function__` for `__deepcopy__` on FakeTensors because the original Tensors were not FakeTensors to begin with. 
- Deepcopying the module and then converting each Tensor attribute to FakeTensor manually. This doesn't set up the right mapping from original Tensor -> new Fake Tensor in case any of those Tensor attributes are used elsewhere.
- Maybe in the future making `FakeTensorMode` also inherit from `TorchFunctionMode` and defining a `__torch__function__` for deepcopy might work, but at present there isnt infrastructure for inheriting from both TorchFunction and TorchDispatch, and when I tried a trivial version of overriding deepcopy with a mode I ran into errors.

The changes here are minimal and we can switch to the third in the future as needed.

Also updates the dynamic shape checking to use tags.


[ghstack-poisoned]

eellison committed:
Update on "Add Fake Tensor deepcopy impl"


So, the goal here is to take an `nn.Module` and convert all of its tensors to `FakeTensors`, and with the correct mapping from the original Tensors to their new FakeTensors which is needed for consistent storages/aliasing relationships in the new FakeTensors. 

A couple things that don't work:
- defining `__torch_function__` for `__deepcopy__` on FakeTensors because the original Tensors were not FakeTensors to begin with. 
- Deepcopying the module and then converting each Tensor attribute to FakeTensor manually. This doesn't set up the right mapping from original Tensor -> new Fake Tensor in case any of those Tensor attributes are used elsewhere.
- Maybe in the future making `FakeTensorMode` also inherit from `TorchFunctionMode` and defining a `__torch__function__` for deepcopy might work, but at present there isnt infrastructure for inheriting from both TorchFunction and TorchDispatch, and when I tried a trivial version of overriding deepcopy with a mode I ran into errors.


Edit: no longer accepting non-fake inputs, and introducing a TorchFunctionMode just for `__deepcopy__`, and invoking it in torch.Tensor/torch.nn.Parameter

The changes here are minimal and we can switch to the third in the future as needed.

Also updates the dynamic shape checking to use tags.


[ghstack-poisoned]

eellison committed:
Update on "Add Fake Tensor deepcopy impl"


So, the goal here is to take an `nn.Module` and convert all of its tensors to `FakeTensors`, and with the correct mapping from the original Tensors to their new FakeTensors which is needed for consistent storages/aliasing relationships in the new FakeTensors. 

A couple things that don't work:
- defining `__torch_function__` for `__deepcopy__` on FakeTensors because the original Tensors were not FakeTensors to begin with. 
- Deepcopying the module and then converting each Tensor attribute to FakeTensor manually. This doesn't set up the right mapping from original Tensor -> new Fake Tensor in case any of those Tensor attributes are used elsewhere.
- Maybe in the future making `FakeTensorMode` also inherit from `TorchFunctionMode` and defining a `__torch__function__` for deepcopy might work, but at present there isnt infrastructure for inheriting from both TorchFunction and TorchDispatch, and when I tried a trivial version of overriding deepcopy with a mode I ran into errors.


Edit: no longer accepting non-fake inputs, and introducing a TorchFunctionMode just for `__deepcopy__`, and invoking it in torch.Tensor/torch.nn.Parameter

The changes here are minimal and we can switch to the third in the future as needed.

Also updates the dynamic shape checking to use tags.


[ghstack-poisoned]

albanD commented on torch/nn/parameter.py:
```diff
@@ -52,10 +52,15 @@ def __new__(cls, data=None, requires_grad=True):
         if id(self) in memo:
             return memo[id(self)]
+
+        if _get_torch_function_mode():
```
I don't think that's the way to go there, this should just properly implement the `__torch_function__` handler (similar to the Tensor `__deepcopy__`).
cc @jbschlosser

eellison replied:
How do you mean ?

jbschlosser replied:
I think he means do this instead of referencing modes:
```python
if has_torch_function_unary(self):
    return handle_torch_function(Tensor.__deepcopy__, (self,), self, memo)
```

However, it looks like this is unneeded in the most recent code update?

jbschlosser replied:
TODO for us to fix `Parameter.__deepcopy__()` to call `self.data.__deepcopy__()` instead of `self.data.clone()`. The two implementations have different view-preserving behavior because of how they interact with the memo.

eellison committed:
Update on "Add Fake Tensor deepcopy impl"


So, the goal here is to take an `nn.Module` and convert all of its tensors to `FakeTensors`, and with the correct mapping from the original Tensors to their new FakeTensors which is needed for consistent storages/aliasing relationships in the new FakeTensors. 

A couple things that don't work:
- defining `__torch_function__` for `__deepcopy__` on FakeTensors because the original Tensors were not FakeTensors to begin with. 
- Deepcopying the module and then converting each Tensor attribute to FakeTensor manually. This doesn't set up the right mapping from original Tensor -> new Fake Tensor in case any of those Tensor attributes are used elsewhere.
- Maybe in the future making `FakeTensorMode` also inherit from `TorchFunctionMode` and defining a `__torch__function__` for deepcopy might work, but at present there isnt infrastructure for inheriting from both TorchFunction and TorchDispatch, and when I tried a trivial version of overriding deepcopy with a mode I ran into errors.


Edit: no longer accepting non-fake inputs, and introducing a TorchFunctionMode just for `__deepcopy__`, and invoking it in torch.Tensor/torch.nn.Parameter

The changes here are minimal and we can switch to the third in the future as needed.

Also updates the dynamic shape checking to use tags.


[ghstack-poisoned]

eellison committed:
Update on "Add Fake Tensor deepcopy impl"


So, the goal here is to take an `nn.Module` and convert all of its tensors to `FakeTensors`, and with the correct mapping from the original Tensors to their new FakeTensors which is needed for consistent storages/aliasing relationships in the new FakeTensors. 

A couple things that don't work:
- defining `__torch_function__` for `__deepcopy__` on FakeTensors because the original Tensors were not FakeTensors to begin with. 
- Deepcopying the module and then converting each Tensor attribute to FakeTensor manually. This doesn't set up the right mapping from original Tensor -> new Fake Tensor in case any of those Tensor attributes are used elsewhere.
- Maybe in the future making `FakeTensorMode` also inherit from `TorchFunctionMode` and defining a `__torch__function__` for deepcopy might work, but at present there isnt infrastructure for inheriting from both TorchFunction and TorchDispatch, and when I tried a trivial version of overriding deepcopy with a mode I ran into errors.


Edit: no longer accepting non-fake inputs, and introducing a TorchFunctionMode just for `__deepcopy__`, and invoking it in torch.Tensor/torch.nn.Parameter

The changes here are minimal and we can switch to the third in the future as needed.

Also updates the dynamic shape checking to use tags.


[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/native/Onehot.cpp:
```diff
@@ -16,7 +16,8 @@ Tensor one_hot(const Tensor &self, int64_t num_classes) {
     }
-    if (self.is_meta()) {
+
+    if (self.key_set().has_all(DispatchKeySet(BackendComponent::MetaBit))) {
```
does calling `is_meta()` directly break something?

eellison replied:
Hey, more context is here: https://github.com/pytorch/pytorch/pull/79170 and https://github.com/pytorch/pytorch/pull/78972

bdhirsh replied:
Ah thanks for the link. Would it be useful if `FakeTensor` were able to treat `one_hot` as a primitive, instead of decomposing it?

That's actually something we want to enable for tensor subclasses soon (internal link: https://fb.quip.com/IlhcA0x3wXqu)

bdhirsh replied:
Ah thanks for the link. Would it be useful if `FakeTensor` were able to treat `one_hot` as a primitive, instead of decomposing it?

That's actually something we want to enable for tensor subclasses soon (internal link: https://fb.quip.com/IlhcA0x3wXqu)

eellison replied:
Yea, that would be helpful!

samdow commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -497,3 +479,28 @@ def to_cpu(e):
+            if id(tensor) in memo:
+                return memo[id(tensor)]
+
+            return self.fake_mode.from_tensor(tensor)
```
Do we want to memoize this?

[See parameter for ref](https://github.com/pytorch/pytorch/blob/master/torch/nn/parameter.py#L57)

eellison replied:
Is 
```
            if id(tensor) in memo:
                return memo[id(tensor)]
```

isn't that memoizing it ?

samdow replied:
Based on parameter, we also need to explicitly write the output to memo if we haven't seen it before

eellison replied:
I think the writing of the `memo` happens here: https://github.com/pytorch/pytorch/blob/6f9b5208dd271579fa0fe76d2a71181937697a4e/torch/nn/parameter.py#L56 right ?

samdow replied:
In parameter, yes. To me this suggests that if we're hitting deepcopy in fake tensor (meaning that we aren't deep copying Parameter(FakeTensor(...)) but just FakeTensor(...)), then we also need to have that same logic in the fake tensor so that memo gets populated

```suggestion
            result = self.fake_mode.from_tensor(tensor)
            memo[id(self)] = result
            return result
```

eellison replied:
and `self.fake_mode.from_tensor` already memoizes conversion from Normal Tensor -> Fake.. 

But I guess yea 
```           
 if id(tensor) in memo:
                return memo[id(tensor)]
``` is a no-op here since we're not writing to it.

samdow replied:
Ah I see. Thanks!

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -497,3 +479,28 @@ def to_cpu(e):
+        # clone will get called in Parameter deepcopy
+        if func == torch._C._TensorBase.clone:
+            return func(self.fake_mode.from_tensor(args[0]), **kwargs)
+        elif func == torch.Tensor.__deepcopy__:
```
This is being tested in `test_deepcopy`, you do hit it if a mode is set

eellison replied:
This is being tested in `BatchNorm`, because it has non-Parameter tensors I believe, and in `ModuleNew` which doesnt use parameters

eellison committed:
Update on "Add Fake Tensor deepcopy impl"


So, the goal here is to take an `nn.Module` and convert all of its tensors to `FakeTensors`, and with the correct mapping from the original Tensors to their new FakeTensors which is needed for consistent storages/aliasing relationships in the new FakeTensors. 

A couple things that don't work:
- defining `__torch_function__` for `__deepcopy__` on FakeTensors because the original Tensors were not FakeTensors to begin with. 
- Deepcopying the module and then converting each Tensor attribute to FakeTensor manually. This doesn't set up the right mapping from original Tensor -> new Fake Tensor in case any of those Tensor attributes are used elsewhere.
- Maybe in the future making `FakeTensorMode` also inherit from `TorchFunctionMode` and defining a `__torch__function__` for deepcopy might work, but at present there isnt infrastructure for inheriting from both TorchFunction and TorchDispatch, and when I tried a trivial version of overriding deepcopy with a mode I ran into errors.


Edit: no longer accepting non-fake inputs, and introducing a TorchFunctionMode just for `__deepcopy__`, and invoking it in torch.Tensor/torch.nn.Parameter

The changes here are minimal and we can switch to the third in the future as needed.

Also updates the dynamic shape checking to use tags.


[ghstack-poisoned]

eellison committed:
Update on "Add Fake Tensor deepcopy impl"


So, the goal here is to take an `nn.Module` and convert all of its tensors to `FakeTensors`, and with the correct mapping from the original Tensors to their new FakeTensors which is needed for consistent storages/aliasing relationships in the new FakeTensors. 

A couple things that don't work:
- defining `__torch_function__` for `__deepcopy__` on FakeTensors because the original Tensors were not FakeTensors to begin with. 
- Deepcopying the module and then converting each Tensor attribute to FakeTensor manually. This doesn't set up the right mapping from original Tensor -> new Fake Tensor in case any of those Tensor attributes are used elsewhere.
- Maybe in the future making `FakeTensorMode` also inherit from `TorchFunctionMode` and defining a `__torch__function__` for deepcopy might work, but at present there isnt infrastructure for inheriting from both TorchFunction and TorchDispatch, and when I tried a trivial version of overriding deepcopy with a mode I ran into errors.


Edit: no longer accepting non-fake inputs, and introducing a TorchFunctionMode just for `__deepcopy__`, and invoking it in torch.Tensor/torch.nn.Parameter

The changes here are minimal and we can switch to the third in the future as needed.

Also updates the dynamic shape checking to use tags.


[ghstack-poisoned]

samdow reviewed (approved):
Thanks for rolling with the changes and exploration!

jbschlosser commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -497,3 +479,30 @@ def to_cpu(e):
+    def __torch_function__(self, func, types, args=(), kwargs=None):
+        kwargs = kwargs if kwargs else {}
+
+        # clone will get called in Parameter deepcopy
```
FYI if we fix #79788, `Parameter.__deepcopy__()` will no longer call `clone()`.

eellison commented:
@pytorchbot merge
